### PR TITLE
Pool class refinements

### PR DESCRIPTION
**Pull request by gregw**
Some updates to the new Pool class:

 + fixed a race with pending reservations
 + use a pending counter
 + Reservation API to simplify Entry API
 + removed public methods on Entry API

The pending counter is a little irksome as I didn't want to grab the lock on every enable.... see comments.

### CODE REVIEW

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 1]
```diff
@@ -128,35 +130,42 @@ public final void setMaxUsageCount(int maxUsageCount)
...
         {
-            lock.unlock();
+            if (closed)
+                return null;
+
+            int space = maxEntries - sharedList.size();
+            if (space <= 0)
+                return null;
+
+            // The pending count is an AtomicInt that is only ever incremented here with
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Nit: "AtomicInt" -> "AtomicInteger" or it would seem like we have created our own class `AtomicInt`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 2]
```diff
@@ -332,29 +372,105 @@ public String toString()
...
             return pooled;
         }
 
-        public void enable(T pooled)
+        /**
+         * Release the entry.
+         * This is equivalent to calling {@link Pool#release(Pool.Entry)} passing this entry.
+         * @return true if released.
+         */
+        public boolean release()
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:429]: This method is never used, so remove it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 3]
```diff
@@ -332,29 +372,105 @@ public String toString()
...
+        {
+            Objects.requireNonNull(pooled);
+            if (entry.state.getHi() != Integer.MIN_VALUE)
+                throw new IllegalStateException("Open entries cannot be enabled : " + this);
+            entry.pooled = pooled;
+            int usage = acquire ? 1 : 0;
+            if (!entry.state.compareAndSet(Integer.MIN_VALUE, usage, 0, usage))
+            {
+                entry.pooled = null;
+                throw new IllegalStateException("Entry cannot be enabled : " + this);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: A `Reservation` may be concurrently removed while it's being enabled.
This code should be converted to the classic `while (true)` loop and retry in case of failures.
Furthermore, it needs to account for the closed state (`hi==-1`) and not enable if it's already closed.
Which means `enable()` should return a `boolean` that tells whether the enabling could be done.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 4]
```diff
@@ -332,29 +372,105 @@ public String toString()
...
+        public void remove()
+        {
+            Pool.this.remove(entry);
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("%s@%x{%s}",
+                getClass().getName(),
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Use `getClass().getSimpleName()` here.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 5]
```diff
@@ -223,8 +223,8 @@ public void succeeded(Connection connection)
             {
                 if (LOG.isDebugEnabled())
                     LOG.debug("Connection {}/{} creation succeeded {}", pool.size(), pool.getMaxEntries(), connection);
-                adopt(entry, connection);
-                future.complete(null);
+                adopt(reservation, connection);
+                future.complete(null);  // TODO could this now be a future that passes back the connection?
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Remove the TODO, as `ConnectionPool` does not work in this way: by the time we notify the future, the `Connection` could already be stolen by another thread.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 6]
```diff
@@ -246,16 +246,16 @@ protected void proceed()
...
+    private void adopt(Pool<Connection>.Reservation reservation, Connection connection)
     {
         if (!(connection instanceof Attachable))
             throw new IllegalArgumentException("Invalid connection object: " + connection);
         Attachable attachable = (Attachable)connection;
-        attachable.setAttachment(entry);
+        attachable.setAttachment(reservation.getEntry());
         if (LOG.isDebugEnabled())
-            LOG.debug("onCreating {}", entry);
+            LOG.debug("adopt {} {}", reservation.getEntry(), connection);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: I would log the `reservation` here, not `reservation.getEntry()`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 7]
```diff
@@ -215,6 +224,42 @@ public Entry acquire()
...
     }
 
+    /**
+     * Utility method to acquire an entry from the pool,
+     * reserving and creating a new entry if necessary.
+     *
+     * @param creator a function to create the pooled value for a reserved entry.
+     * @return an entry from the pool or null if none is available.
+     */
+    public Entry acquire(Function<Entry, T> creator)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: There are no tests for this new method.

Furthermore, why does it take an `Entry` as parameter? I think it should just be a `Supplier<T>` to create the object as the creator is called within the acquisition process, so the implementation does all the plumbing with the `Entry`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 8]
```diff
@@ -281,17 +326,12 @@ public boolean isClosed()
...
             copy = new ArrayList<>(sharedList);
             sharedList.clear();
         }
-        finally
-        {
-            lock.unlock();
-        }
 
         // iterate the copy and close its entries
         for (Entry entry : copy)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:340]: Can we use `IO.close()` in the loop? I think a `warn()` here is too much.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 9]
```diff
@@ -215,6 +224,42 @@ public Entry acquire()
...
+            return null;
+
+        T value = null;
+        try
+        {
+            value = creator.apply(reservation.getEntry());
+        }
+        catch (Throwable th)
+        {
+            LOGGER.warn(th);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: I think we should call `reservation.remove()` and rethrow here - same semantic of `Map.computeIfAbsent()`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 10]
```diff
@@ -215,6 +224,42 @@ public Entry acquire()
...
+        }
+
+        if (value == null)
+        {
+            reservation.remove();
+            return null;
+        }
+
+        return reservation.acquire(value);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: It may be possible that the pool is closed just before the call to this method, so we should return `null`, which means `reservation.acquire()` should return a `boolean`.

**Review by sbordet — CHANGES_REQUESTED** [review]
Now that MIN_VALUE indicates "pending", in many places the logic `boolean closed = state.getHi() < 0` is now imprecise.
This happens in `tryAcquire()`, `tryRelease()`, `isClosed()`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 7 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: It takes an entry because it is typical (both our usages) for a two way association to be made between the pooled item and its entry.  This allows the item to pass the entry in the constructor or associate it with an attachment etc.

I'll add some tests

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 9 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: just throw and not warn?   I don't really like log and throw.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 10 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Currently it throws an ISE, which is what enable did before.  reservation.acquire can't return a boolean because it returns the entry.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 3 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Unfortunately this code is just not threadsafe and I can't see how we can make it so without adding a lock.
The problem is that if we allow multiple threads to call enable on the reservation, both may pass the MIN_VALUE check and then both will try to set the `entry.pooled` value.  Only one will win and it may not be the one that wins the CaS race.

So currently enable will ISE atomically if anybody else messes with state (of the reserved entry or by closing the pool).  The state of the pooled field is not so well defined for multiple threads.... but I think any app that is racing itself to call enable is just wrong anyway.

I think ISE is OK for -1 state.... but are you saying that if it is -1, we should noop?       If so, I can do that after a failed CaS, but not in a loop, which will never work because of the pooled field problem.

I've gone back on forth on this method a lot... I don't like it, but I like it more than a big lock.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 2 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:429]: It will be once I do my XmlConfiguration PR.
It makes sense that an item in the pool that has a reference to it's own entry can release itself without needing to keep a reference to the pool.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 8 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:340]: Sure.... but should we close any Closeable when it is removed from the pool, or only when the pool is closed?  Perhaps we should close AutoCloseables that are removed from the pool.  Eitherway, need to javadoc this behavior

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 7 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: I think the two way association with `Attachable` should be done in this method, and not be a burden of the implementer of the creator lambda.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 10 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: No, this last `return` statement should be:
```
if (reservation.acquire(value))
  return reservation.getEntry();
else
  return null;
```

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 3 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: > The problem is that if we allow multiple threads to call enable on the reservation

But we never supported that and I think it's ok - `enable()` has a natural race with `remove()` but not with itself.

It is possible to write the classic `while (true)` loop, not care about races with itself, but do care about races with `remove()` that are natural and should not throw.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 2 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:429]: I'm not sure. `Entry` is not really an API class now that you have introduced `Reservation`.
It's an opaque mechanism that needs to be exposed, but I think it's more correct to force people to use `Pool` and `Reservation`and `Attachable` rather than `Entry`.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by lorban — CHANGES_REQUESTED** [review]
I see three different changes in this PR:
 - general small cleanups like removing `public` visibility on the 3 `Entry.try*` methods.
 - the introduction of a new `pending` atomic which requires the introduction of a new Entry state.
 - the introduction of the `Reservation` class and `acquire(Function)` method.

The first item is mostly good, I have few objections.
I agree on principle on the second item, but it would need some extra work and cleanup like a new `isReserved()` method and some more tests.
The third one I absolutely don't like it. It introduces higher-level constructs that IMHO should not be in the `Pool` class as it would blur its API: both low-level and high(er) level operations would be presented which would add confusion over time. I do see some value in your changes but they should go to another class that would wrap the pool and only expose the higher-level API you wanna use.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 7 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: I don't agree.  There is currently no dependency on `Attachable` from Pool and nothing that says an `Attachable` must be used for the association... or even if something is `Attachable` that it may be so for another reason.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 10 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: But the acquire will throw ISE if the pool has been closed... just as it did before.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 3 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: The class `while (true)` loop doesn't work if we throw ISE if the state is not what we expected.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 2 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:429]: `Entry` most definitely is an API class and I have code that pools objects that remember their own `Entry`.  It is pointless to force them to also remember the pool they came from just so they can call release.... as the `Entry` already knows its own pool.

**Comment by gregw** [comment]
@lorban Can you give me actual feedback in the code rather than just general impressions, as I don't really understand your "objections" if they are just described as "objectons".

Note that now I know this class better, I very much am in favour of something like the `Reservation` mechanism because:
 + it is wrong for the Entry method to carry a mix of lifecycle methods and long lived methods.
 + Keeping the pending count correct depends entirely on the correct implementation of the lifecycle contract of a Reservation
 + I think it is important that the class returned from `acquire()` has no lifecycle methods on it other than the matching `release()`.  It is only callers of `reserve(...)` that should receive a class with access to life cycle methods.

So I'm very sure that we need to fix that interface.   I'm not so sure that the `Reservation` is the only way to do it, but it's my proposal and I need to know what your specific objections are to it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 7 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Note also that with the `Reservation` approach, this method is not so important as the API is clear and it is probably less faff to not use this method.   So if we don't really agree on this method, it is OK to just not have it... so long as we have `Reservation` or something like it.

**Comment by lorban** [comment]
@gregw I like your idea about the `Reservation` mechanism too, but IMHO its place isn't in the `Pool` class. Why? Because you're conflating object factorizing with pooling which have opposing goals, your discussion with @sbordet about `Attachable` and `ISE` reflects this.

In my view, we'd better not touch `Pool` at all (or marginally) and create a different class reconciling the factorizing and pooling concepts so here's my proposal:

```java
public class PoolingFactory<T>
{
    private final Pool<T> _pool;
    private final Function<Reservation, T> _creator;

    public PoolingFactory(Function<Reservation, T> creator, int maxPoolSize, int cacheSize)
    {
        _creator = creator;
        _pool = new Pool<>(maxPoolSize,cacheSize);
    }

    public T get()
    {
        T result;

        Pool<T>.Entry entry = _pool.acquire();
        if (entry != null)
        {
            result = entry.getPooled();
        }
        else
        {
            entry = _pool.reserve(-1);
            result = _creator.apply(new Reservation(entry));
            if (entry != null)
                entry.enableAcquired(result); // this is a new, easy to implement method
        }

        return result;
    }

    public class Reservation implements AutoCloseable
    {
        private final Pool<T>.Entry entry;

        public Reservation(Pool<T>.Entry entry)
        {
            this.entry = entry;
        }

        @Override
        public void close()
        {
            if (entry != null)
                _pool.release(entry);
        }
    }
}
```

and applying that to the `ConfigurationParser` would look like this:

```java
    private static class ConfigurationParser extends XmlParser implements AutoCloseable
    {
        private final PoolingFactory<ConfigurationParser>.Reservation _reservation;

        private ConfigurationParser(PoolingFactory<ConfigurationParser>.Reservation reservation)
        {
            _reservation = Objects.requireNonNull(reservation);
            ...
        }

        @Override
        public void close()
        {
            _reservation.close();
        }
    }

```

while using it would look like that:

```java
    private static final PoolingFactory<ConfigurationParser> __parsers = new PoolingFactory<>(ConfigurationParser::new, Runtime.getRuntime().availableProcessors(),1);

    public void parseSomething()
    {
        try (ConfigurationParser parser = __parsers.get())
        {
            parser.parse(...);
        }
    }

```

The class name makes it clear that both factorizing and pooling is performed, and its contract is trivial: there's only one method to pool/build, the `Reservation` is closeable and always non-null (even for non-pooled objects) and the construction simply takes a non-capturing lambda onto the pooled/built class' constructor.

So we end up with a different class only exposing a minimal API very specific to this build-and-pool mechanism that is easy both to use and to maintain because the two orthogonal concepts are kept clearly separated.

**Comment by gregw** [comment]
@lorban I think you are confusing the method with a creator with the need for the `Reservation`.  

By adding `Reservation` to Pool mostly removes the need for a creator based method like create, because the API to pool now will not allow it to be used incorrectly.
So let's hangout when we get a chance as I really REALLY think Pool's API needs to have something like Reservation so that it cannot be used incorrectly.

I'm in and out today, so let's try tomorrow.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 7 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: With the Reservation API, it is no longer difficult to call the API directly, so this method is no longer needed.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 11]
```diff
@@ -215,6 +225,43 @@ public Entry acquire()
...
+     * @param creator a function to create the pooled value for a reserved entry.
+     * @return an entry from the pool or null if none is available.
+     */
+    public Entry acquire(Function<Pool<T>.Entry, T> creator)
+    {
+        Entry entry = acquire();
+        if (entry != null)
+            return entry;
+
+        entry = reserve(getMaxEntries());
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Minor nit: you probably should call `reserve(-1)` to make it more obvious that you do not want to be limited by the pending/disabled entries count.

File: jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java [hunk 12]
```diff
@@ -31,16 +35,33 @@
...
+        data.add(new Object[]{1});
+        data.add(new Object[]{2});
+        return data.stream();
+    }
 
-    @Test
-    public void testAcquireRelease()
+    @ParameterizedTest
+    @MethodSource(value = "cacheSize")
+    public void testCache(int cacheSize)
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:54]: This test is useless.

File: jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java [hunk 13]
```diff
@@ -350,35 +419,14 @@ public void testMultiplexRemoveAfterAcquire()
         assertThat(pool.size(), is(0));
     }
 
-    @Test
-    public void testReleaseThenRemoveNonEnabledEntry()
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:354]: Why did you remove this test?

File: jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java [hunk 14]
```diff
@@ -350,35 +419,14 @@ public void testMultiplexRemoveAfterAcquire()
...
-        Pool<String>.Entry e = pool.reserve(-1);
-        assertThat(pool.size(), is(1));
-        assertThat(pool.release(e), is(false));
-        assertThat(pool.size(), is(1));
-        assertThat(pool.remove(e), is(true));
-        assertThat(pool.size(), is(0));
-    }
-
-    @Test
-    public void testRemoveNonEnabledEntry()
```
Comment (lorban) [inline jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:366]: Why did you remove this test?

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 15]
```diff
@@ -215,6 +225,43 @@ public Entry acquire()
...
+            return null;
+
+        T value = null;
+        try
+        {
+            value = creator.apply(entry);
+        }
+        catch (Throwable th)
+        {
+            Pool.this.remove(entry);
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: I don't think the `Pool.this.` prefix is needed.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 16]
```diff
@@ -215,6 +225,43 @@ public Entry acquire()
...
+            throw th;
+        }
+
+        if (value == null)
+        {
+            Pool.this.remove(entry);
```
Comment (lorban) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: I don't think the `Pool.this.` prefix is needed.

**Review by lorban — CHANGES_REQUESTED** [review]
Just a few minor things, otherwise LGTM.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 17]
```diff
@@ -201,29 +201,30 @@ protected void tryCreate(int maxPending)
...
         destination.newConnection(new Promise<Connection>()
         {
             @Override
             public void succeeded(Connection connection)
             {
                 if (LOG.isDebugEnabled())
                     LOG.debug("Connection {}/{} creation succeeded {}", pool.size(), pool.getMaxEntries(), connection);
-                adopt(reservation, connection);
+                if (!(connection instanceof Attachable))
+                    throw new IllegalArgumentException("Invalid connection object: " + connection);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: No, you can't throw from a `Callback` - good that `adopt()` was inlined because I missed it.
Call `failed(new IllegalArgumentException())`, as we need to remove the `Entry`, notify the CF and the requester.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 18]
```diff
@@ -417,6 +381,40 @@ public String toString()
             this.state = new AtomicBiInteger(Integer.MIN_VALUE, 0);
         }
 
+        /** Enable a reserved entry {@link Entry}.
+         * An entry returned from the {@link #reserve(int)} method must be enabled with this method,
+         * once and only once, before it is usable by the pool.
+         * The entry may be enabled and not acquired, in which case it is immediately available to be
+         * acquired, potentially by another thread; or it can be enabled and acquired atomically so that
+         * no other thread can acquired it, although the acquire may still fail if the pool has been closed.
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Typo: acquired -> acquire.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by lorban — APPROVED** [review]
LGTM

**Review by sbordet — APPROVED** [review]
